### PR TITLE
Add rpm packing

### LINKED ISSUE #149: Add rpm type of packing in tt pack
`tt pack` is supposed to be able to pack also rpm packages.

### PR DESCRIPTION
**Pull request by vr009**
With this patch:
- added rpm pacakging
- added according integration tests

Also:
- replaced useless copyBundle function with copy function from otiai10/copy package 
- reused functions from util package for handling versions instead of using self-written functions
- the code for receiving the package name moved to a separate function
- fixed the format of log output
- fixed the copying of binaries logic

Close #149 
Follow up #18

### CODE REVIEW

File: cli/pack/common.go [hunk 1]
```diff
@@ -534,3 +534,29 @@ func copyBinaries(srcPath, destPath string) error {
...
+
+// getPackageName returns the result name of the package.
+func getPackageName(packCtx *cmdcontext.PackCtx, suffix string, addVersion bool) (string, error) {
+	var tarName, versionSuffix string
+	if addVersion {
+		versionSuffix = getVersion(packCtx)
+	}
+
+	if packCtx.FileName != "" {
+		return packCtx.FileName, nil
```
Comment (psergee) [inline cli/pack/common.go:546]: Looks like, this must be checked first. Before `if addVersion`

File: cli/pack/common.go [hunk 2]
```diff
@@ -534,3 +534,29 @@ func copyBinaries(srcPath, destPath string) error {
...
+	} else if packCtx.Name != "" {
+		tarName = packCtx.Name
```
Comment (psergee) [inline cli/pack/common.go:548]: Suppose, this is not necessary tar name. Probably packageName.

File: cli/pack/cpio.go [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"os"
+	"os/exec"
+	"strings"
+)
+
+func packCpio(relPaths []string, resFileName, packageFilesDir string) error {
```
Comment (psergee) [inline cli/pack/cpio.go:12]: pls, add comments

File: cli/pack/cpio.go [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+	filesBuffer := bytes.Buffer{}
+	filesBuffer.WriteString(strings.Join(relPaths, "\n"))
+
+	cpioFile, err := os.Create(resFileName)
+	if err != nil {
```
Comment (psergee) [inline cli/pack/cpio.go:17]: Why do we fill filesBuffer before this check?

File: cli/pack/deb.go [hunk 5]
```diff
@@ -67,65 +67,64 @@ func (packer *debPacker) Run(cmdCtx *cmdcontext.CmdCtx) error {
 	}
 	defer os.RemoveAll(bundlePath)
 
+	log.Infof("Creating a data directory")
```
Comment (psergee) [inline cli/pack/deb.go:70]:
```suggestion
	log.Info("Creating a data directory")
```

File: cli/pack/deb.go [hunk 6]
```diff
@@ -67,65 +67,64 @@ func (packer *debPacker) Run(cmdCtx *cmdcontext.CmdCtx) error {
...
+
 	// Create a data directory.
 	rootPrefixDir := dataDirName
 	rootPrefix := filepath.Join(rootPrefixDir, debBundlePath)
-	if packCtx.Name != "" {
-		rootPrefix = fmt.Sprintf(rootPrefix, packCtx.Name)
-	} else {
-		rootPrefix = fmt.Sprintf(rootPrefix, "bundle")
+
+	packageName, err := getPackageName(&cmdCtx.Pack, "", false)
```
Comment (psergee) [inline cli/pack/deb.go:76]: Why is there not .deb suffix?

File: cli/pack/deb.go [hunk 7]
```diff
@@ -67,65 +67,64 @@ func (packer *debPacker) Run(cmdCtx *cmdcontext.CmdCtx) error {
...
-	log.Infof("Create data tgz")
-
 	// Create data.tar.gz.
 	dataArchivePath := filepath.Join(packageDir, dataArchiveName)
 	err = WriteTgzArchive(filepath.Join(packageDir, rootPrefixDir), dataArchivePath)
 	if err != nil {
 		return err
 	}
 
+	log.Infof("Creating a control directory")
```
Comment (psergee) [inline cli/pack/deb.go:101]:
```suggestion
	log.Info("Creating a control directory")
```

File: cli/pack/deb.go [hunk 8]
```diff
@@ -67,65 +67,64 @@ func (packer *debPacker) Run(cmdCtx *cmdcontext.CmdCtx) error {
...
-
 	// Create control.tar.gz.
 	controlArchivePath := filepath.Join(packageDir, controlArchiveName)
 	err = WriteTgzArchive(controlDirPath, controlArchivePath)
 	if err != nil {
 		return err
 	}
 
-	log.Debugf("Create debian-binary file")
+	log.Infof("Creating a debian-binary file")
```
Comment (psergee) [inline cli/pack/deb.go:117]:
```suggestion
	log.Info("Creating a debian-binary file")
```

File: cli/pack/deb.go [hunk 9]
```diff
@@ -160,13 +160,8 @@ func createDebianBinary(packageDir string) error {
...
 
 // getTntTTVersions checks the version of tarantool in bin_dir and returns it.
-func getTntTTVersions(packCtx *cmdcontext.PackCtx) (PackDependencies, error) {
-	tntVerBytes, err := exec.Command(filepath.Join(packCtx.App.BinDir, "tarantool"), "--version").
-		Output()
-	if err != nil {
-		return nil, err
-	}
-	tntVer, err := util.NormalizeGitVersion(string(tntVerBytes))
+func getTntTTVersions(cmdCtx *cmdcontext.CmdCtx) (PackDependencies, error) {
```
Comment (psergee) [inline cli/pack/deb.go:163]: If this function returns Tarantool version, why is "TT" in its name? Why Versions (plural form)?

File: cli/pack/deb_control_dir.go [hunk 10]
```diff
@@ -34,8 +34,11 @@ type PackDependency struct {
 type PackDependencies []PackDependency
 
 // createControlDir creates a control directory that contains control file, postinst and preinst.
-func createControlDir(packCtx *cmdcontext.PackCtx, destDirPath string) error {
+func createControlDir(cmdCtx *cmdcontext.CmdCtx, destDirPath string) error {
+	packCtx := &cmdCtx.Pack
+
 	log.Debugf("Create DEB control file")
```
Comment (psergee) [inline cli/pack/deb_control_dir.go:40]:
```suggestion
	log.Debug("Create DEB control file")
```

File: cli/pack/deb_control_dir.go [hunk 11]
```diff
@@ -83,7 +86,8 @@ func createControlDir(packCtx *cmdcontext.PackCtx, destDirPath string) error {
 	if err != nil {
 		return err
 	}
-	log.Infof("Created control in %s", destDirPath)
+
+	log.Debugf("Created control in %s", destDirPath)
```
Comment (psergee) [inline cli/pack/deb_control_dir.go:90]:
```suggestion
	log.Debugf("Created control file in %s", destDirPath)
```

File: cli/pack/rpm.go [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+	"os"
+	"path/filepath"
+)
+
+// rpmPacker is a structure that implements Packer interface
+// with specific rpm packing behavior.
+type rpmPacker struct {
+}
+
+// Run packs a bundle into deb package.
```
Comment (psergee) [inline cli/pack/rpm.go:18]: Why deb?

File: cli/pack/common.go (hunk 1 above)
Comment (vr009) [reply cli/pack/common.go:544]: Right, fixed.

File: cli/pack/common.go (hunk 2 above)
Comment (vr009) [reply cli/pack/common.go]: Oops sorry, the code was moved from tar-specific code, fixed now.

File: cli/pack/cpio.go (hunk 4 above)
Comment (vr009) [reply cli/pack/cpio.go:15]: Fixed. Sorry, the code is taken from cartridge-cli implementation.

File: cli/pack/deb.go (hunk 6 above)
Comment (vr009) [reply cli/pack/deb.go:77]: I need a non suffixed name here for creating a directory in /usr/share/tarantool with this name, where the installed package will be located.

File: cli/pack/deb.go (hunk 7 above)
Comment (vr009) [reply cli/pack/deb.go]: Fixed.

File: cli/pack/deb.go (hunk 8 above)
Comment (vr009) [reply cli/pack/deb.go]: Fixed.

File: cli/pack/deb_control_dir.go (hunk 10 above)
Comment (vr009) [reply cli/pack/deb_control_dir.go]: Fixed.

File: cli/pack/deb_control_dir.go (hunk 11 above)
Comment (vr009) [reply cli/pack/deb_control_dir.go]: Fixed.

File: cli/pack/rpm.go (hunk 12 above)
Comment (vr009) [reply cli/pack/rpm.go]: Typo, fixed.

File: cli/pack/cpio.go (hunk 3 above)
Comment (vr009) [reply cli/pack/cpio.go:13]: Added.

File: cli/pack/deb.go (hunk 9 above)
Comment (vr009) [reply cli/pack/deb.go:167]: The name is correct at the moment, it returns tt version too now.

**Comment by psergee** [comment]
On my Ubuntu basic functionality does not work:
```
../tt/tt pack rpm --name demo --version 0.1.1 
   ⨯ Incorrect type of package
```

1. 
```
../../tt/tt pack dsf
   ⨯ Incorrect type of package
```
Could you please add supported pack types somewhere? It would be nice to see them in error message. Add integration test check this error case.

2. A have an application in current dir and the folowing tt config:
```
tt:
  modules:
  app:
    instances_enabled: .
```
tt start and stop works fine. But tt pack gives the following error:
```
../../tt/tt pack deb
   ⨯ Failed to pack: open /home/psergee/work/tmp/demo/bin: no such file or directory
```

3. Okay, I made tt pack work, but I see wrong result package path:
```
../../tt/tt pack deb --name demo
     • Created result DEB package: /tmp/tt_pack1737719650
```

4. Some information printed to console during packing looks odd for me. For example:
A root for package is located in: /tmp/1184594429
. . .
Running `/tmp/tt_pack1737719650/myapp/cartridge.pre-build`
Why are these temporary directories are different? What is "root for package" and what is tt_pack1737719650?

`Initialize the app directory for prefix: data/usr/share/tarantool/demo`
This path is relative. But I have no idea what it's relative to.

5.
`../tt/tt pack deb --name demo --version 0.1.1`
creates demo.deb package for me. No version.

File: cli/pack/rpm.go [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+		return err
+	}
+	defer os.RemoveAll(bundlePath)
+
+	bundleName, err := getPackageName(&cmdCtx.Pack, "", false)
+	if err != nil {
+		return err
+	}
+
+	if err := copy.Copy(bundlePath, filepath.Join(packageDir, "usr/share/tarantool",
```
[the remark this answers is not in the document]
Comment (vr009) [reply cli/pack/rpm.go]: Fixed.

File: cli/pack/rpm_header.go [hunk 14]
```diff
@@ -0,0 +1,269 @@
...
+		return nil, fmt.Errorf("Failed to get payload digest: %s", err)
+	}
+
+	cpioFileInfo, err := os.Stat(cpioPath)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to get payload size: %s", err)
+	}
+	payloadSize := cpioFileInfo.Size()
+
+	// gen fileinfo
```
[the remark this answers is not in the document]
Comment (vr009) [reply cli/pack/rpm_header.go]: Fixed.

File: cli/pack/rpm_header.go [hunk 15]
```diff
@@ -0,0 +1,269 @@
...
+			Value: postInst},
+	}...)
+}
+
+// genRpmHeader generates rpm headers.
+func genRpmHeader(relPaths []string, cpioPath, compresedCpioPath, packageFilesDir string,
+	cmdCtx *cmdcontext.CmdCtx) (rpmTagSetType, error) {
+	rpmHeader := rpmTagSetType{}
+
+	// compute payload digest
```
[the remark this answers is not in the document]
Comment (vr009) [reply cli/pack/rpm_header.go]: Fixed.

File: cli/pack/rpm_header.go [hunk 16]
```diff
@@ -0,0 +1,269 @@
...
+		{ID: tagPayloadDigest, Type: rpmTypeStringArray, Value: []string{payloadDigest}},
+		{ID: tagPayloadDigestAlgo, Type: rpmTypeInt32, Value: []int32{int32(payloadDigestAlgo)}},
+	}...)
+
+	deps, err := parseDependencies(cmdCtx.Pack.RpmDeb.Deps)
+	if err != nil {
+		return nil, err
+	}
+	if cmdCtx.Pack.RpmDeb.DepsFile != "" {
+		deps, err = parseDependenciesFromFile(cmdCtx.Pack.RpmDeb.Deps, cmdCtx.Pack.RpmDeb.DepsFile)
```
[the remark this answers is not in the document]
Comment (vr009) [reply cli/pack/rpm_header.go]: Fixed. Also I've moved the logic of fetching dependencies from file, deps-list and tarantool/tt binaries to a separate function.

File: cli/pack/rpm_header.go [hunk 17]
```diff
@@ -0,0 +1,269 @@
...
+
+	for _, relPath := range relPaths {
+		fullFilePath := filepath.Join(dirPath, relPath)
+		fileInfo, err := os.Stat(fullFilePath)
+		if err != nil {
+			return filesInfo, err
+		}
+
+		if fileInfo.Mode().IsRegular() {
+			filesInfo.FileFlags = append(filesInfo.FileFlags, fileFlag) // XXX
```
[the remark this answers is not in the document]
Comment (vr009) [reply cli/pack/rpm_header.go]: I have no idea. The code got taken from cartridge-cli. Removed the comment.

File: cli/pack/rpm_header.go [hunk 18]
```diff
@@ -0,0 +1,269 @@
...
+	"strconv"
+	"strings"
+	"syscall"
+
+	"github.com/tarantool/tt/cli/cmdcontext"
+	"github.com/tarantool/tt/cli/util"
+	"github.com/tarantool/tt/cli/version"
+)
+
+type filesInfoType struct {
```
[the remark this answers is not in the document]
Comment (vr009) [reply cli/pack/rpm_header.go]: Fixed.

File: cli/pack/rpm_pack.go [hunk 19]
```diff
@@ -0,0 +1,218 @@
...
+ *
+ *  Both signature sections have the same format: a set of typed
+ *  key-value pairs.
+ *
+ *  While debugging, I used rpm-dissecting tool from mkrepo:
+ *  - https://github.com/tarantool/mkrepo/blob/master/mkrepo.py
+ *
+ */
+
+// packRpm creates an RPM archive in cmdCtx.Pack.ResPackagePath
```
[the remark this answers is not in the document]
Comment (vr009) [reply cli/pack/rpm_pack.go]: The comment was taken from cartridge-cli and was not changed, my fault. There are no `cmdCtx.Pack.ResPackagePath` and `cmdCtx.Pack.PackageFilesDir`.

File: cli/pack/rpm_pack.go [hunk 20]
```diff
@@ -0,0 +1,218 @@
...
+// that contains files from cmdCtx.Pack.PackageFilesDir.
+func packRpm(cmdCtx *cmdcontext.CmdCtx, packageDir, resPackagePath string) error {
+	var err error
+
+	relPaths, err := getSortedRelPaths(packageDir)
+	if err != nil {
+		return fmt.Errorf("Failed to get sorted package files list: %s", err)
+	}
+
+	log.Infof("Creating data section")
```
[the remark this answers is not in the document]
Comment (vr009) [reply cli/pack/rpm_pack.go]: Fixed.

File: cli/pack/rpm_pack.go [hunk 21]
```diff
@@ -0,0 +1,218 @@
...
+	if err := packCpio(relPaths, cpioPath, packageDir); err != nil {
+		return fmt.Errorf("Failed to pack CPIO: %s", err)
+	}
+
+	compresedCpioPath := filepath.Join(packageDir, "cpio.gz")
+	if err := CompressGzip(cpioPath, compresedCpioPath); err != nil {
+		return fmt.Errorf("Failed to compress CPIO: %s", err)
+	}
+
+	log.Infof("Generating header section")
```
[the remark this answers is not in the document]
Comment (vr009) [reply cli/pack/rpm_pack.go]: Fixed.

File: cli/pack/rpm_pack.go [hunk 22]
```diff
@@ -0,0 +1,218 @@
...
+	if err != nil {
+		return fmt.Errorf("Failed to create RPM body file: %s", err)
+	}
+	defer rpmHeaderFile.Close()
+
+	if _, err := io.Copy(rpmHeaderFile, packedHeader); err != nil {
+		return fmt.Errorf("Failed to write RPM lead to file: %s", err)
+	}
+
+	// create body file = header + compressedCpio
```
[the remark this answers is not in the document]
Comment (vr009) [reply cli/pack/rpm_pack.go]: Fixed.

File: cli/pack/rpm_pack.go [hunk 23]
```diff
@@ -0,0 +1,218 @@
...
+	rpmBodyFilePath := filepath.Join(packageDir, "body")
+	if err := util.MergeFiles(rpmBodyFilePath, rpmHeaderFilePath, compresedCpioPath); err != nil {
+		return fmt.Errorf("Failed to concat RPM header with compressed payload: %s", err)
+	}
+
+	log.Infof("Computing a signature")
```
[the remark this answers is not in the document]
Comment (vr009) [reply cli/pack/rpm_pack.go]: Fixed.

File: cli/util/util.go [hunk 24]
```diff
@@ -746,3 +747,39 @@ func GetAbsPath(baseDir, filePath string) string {
...
+	destFile, err := os.Create(destFilePath)
+	if err != nil {
+		return fmt.Errorf("Failed to create result file %s: %s", destFilePath, err)
+	}
+	defer destFile.Close()
+
+	for _, srcFilePath := range srcFilePaths {
+		srcFile, err := os.Open(srcFilePath)
+		if err != nil {
+			return fmt.Errorf("Failed to open source file %s: %s", srcFilePath, err)
```
[the remark this answers is not in the document]
Comment (vr009) [reply cli/util/util.go:775]: Fixed.

File: cli/util/util.go [hunk 25]
```diff
@@ -746,3 +747,39 @@ func GetAbsPath(baseDir, filePath string) string {
...
+		}
+
+		_, err = io.Copy(destFile, srcFile)
+		srcFile.Close()
+
+		if err != nil {
+			return err
```
[the remark this answers is not in the document]
Comment (vr009) [reply cli/util/util.go:782]: Fixed.

File: cli/pack/rpm_pack.go [hunk 26]
```diff
@@ -0,0 +1,218 @@
...
+		return fmt.Errorf("Failed to gen RPM signature: %s", err)
+	}
+
+	packedSignature, err := packTagSet(*signature, headerSignatures)
+	if err != nil {
+		return fmt.Errorf("Failed to pack RPM header: %s", err)
+	}
+	alignData(packedSignature, 8)
+
+	log.Infof("Computing lead section")
```
[the remark this answers is not in the document]
Comment (vr009) [reply cli/pack/rpm_pack.go]: Fixed.

File: cli/pack/rpm_pack.go [hunk 27]
```diff
@@ -0,0 +1,218 @@
...
+
+	// compute lead
```
[the remark this answers is not in the document]
Comment (vr009) [reply cli/pack/rpm_pack.go]: Fixed.

File: cli/pack/rpm_pack.go [hunk 28]
```diff
@@ -0,0 +1,218 @@
...
+	name, err := getPackageName(&cmdCtx.Pack, "", false)
+	if err != nil {
+		return err
+	}
+	lead := genRpmLead(name)
+	if err := util.ConcatBuffers(lead, packedSignature); err != nil {
+		return err
+	}
+
+	// create lead file
```
[the remark this answers is not in the document]
Comment (vr009) [reply cli/pack/rpm_pack.go]: Fixed.

File: cli/pack/rpm_pack.go [hunk 29]
```diff
@@ -0,0 +1,218 @@
...
+		if err != nil {
+			return err
+		}
+
+		filePath, err = filepath.Rel(srcDir, filePath)
+		if err != nil {
+			return err
+		}
+
+		// system dirs shouldn't be added to the paths list
```
[the remark this answers is not in the document]
Comment (vr009) [reply cli/pack/rpm_pack.go]: Fixed.

File: cli/pack/rpm_tagset.go [hunk 30]
```diff
@@ -0,0 +1,390 @@
...
+
+		packed.Count = len(byteArray)
+		for _, byteValue := range byteArray {
+			if _, err := io.Copy(packed.Data, packValues(byteValue)); err != nil {
+				return nil, err
+			}
+		}
+
+	case rpmTypeStringArray: // STRING_ARRAY
+		// value should be strings array
```
[the remark this answers is not in the document]
Comment (vr009) [reply cli/pack/rpm_tagset.go]: Fixed.

File: cli/pack/rpm_tagset.go [hunk 31]
```diff
@@ -0,0 +1,390 @@
...
+		for _, v := range stringsArray {
+			bytedString := []byte(v)
+			bytedString = append(bytedString, 0)
+			if _, err := io.Copy(packed.Data, packValues(bytedString)); err != nil {
+				return nil, err
+			}
+		}
+
+	case rpmTypeString: // STRING
+		// value should be string
```
[the remark this answers is not in the document]
Comment (vr009) [reply cli/pack/rpm_tagset.go]: Fixed.

File: cli/pack/rpm_tagset.go [hunk 32]
```diff
@@ -0,0 +1,390 @@
...
+		packed.Count = 1
+
+		bytedString := []byte(stringValue)
+		bytedString = append(bytedString, 0)
+		if _, err := io.Copy(packed.Data, packValues(bytedString)); err != nil {
+			return nil, err
+		}
+
+	case rpmTypeInt8: // INT8
+		// value should be []int8
```
[the remark this answers is not in the document]
Comment (vr009) [reply cli/pack/rpm_tagset.go]: Fixed.

File: cli/pack/rpm_tagset.go [hunk 33]
```diff
@@ -0,0 +1,390 @@
...
+		}
+
+		packed.Count = len(int8Values)
+
+		if _, err := io.Copy(packed.Data, packValues(int8Values)); err != nil {
+			return nil, err
+		}
+
+	case rpmTypeInt16: // INT16
+		// value should be []int16
```
[the remark this answers is not in the document]
Comment (vr009) [reply cli/pack/rpm_tagset.go]: Fixed.

File: cli/pack/rpm_tagset.go [hunk 34]
```diff
@@ -0,0 +1,390 @@
...
+		packed.Count = len(int32Values)
+
+		for _, value := range int32Values {
+			if _, err := io.Copy(packed.Data, packValues(value)); err != nil {
+				return nil, err
+			}
+		}
+
+	case rpmTypeInt64: // INT64
+		// value should be []int64
```
[the remark this answers is not in the document]
Comment (vr009) [reply cli/pack/rpm_tagset.go]: Fixed.

File: cli/pack/tgz.go [hunk 35]
```diff
@@ -81,6 +81,39 @@ func WriteTarArchive(srcDirPath string, compressWriter io.Writer) error {
 	return nil
 }
 
+// CompressGzip compresses specified file  with gzip.BestCompression level
```
[the remark this answers is not in the document]
Comment (vr009) [reply cli/pack/tgz.go]: Fixed.

File: cli/pack/tgz.go [hunk 36]
```diff
@@ -81,6 +81,39 @@ func WriteTarArchive(srcDirPath string, compressWriter io.Writer) error {
...
+func CompressGzip(srcFilePath string, destFilePath string) error {
+	var err error
+
+	// src file reader
```
[the remark this answers is not in the document]
Comment (vr009) [reply cli/pack/tgz.go]: Fixed.

File: cli/pack/tgz.go [hunk 37]
```diff
@@ -81,6 +81,39 @@ func WriteTarArchive(srcDirPath string, compressWriter io.Writer) error {
 	return nil
 }
 
+// CompressGzip compresses specified file  with gzip.BestCompression level
+func CompressGzip(srcFilePath string, destFilePath string) error {
+	var err error
```
[the remark this answers is not in the document]
Comment (vr009) [reply cli/pack/tgz.go]: Fixed.

File: cli/pack/tgz.go [hunk 38]
```diff
@@ -81,6 +81,39 @@ func WriteTarArchive(srcDirPath string, compressWriter io.Writer) error {
...
+
+	// src file reader
+	srcFileReader, err := os.Open(srcFilePath)
+	if err != nil {
+		return fmt.Errorf("Failed to open source file %s: %s", srcFilePath, err)
+	}
+	defer srcFileReader.Close()
+
+	// dest file writer
```
[the remark this answers is not in the document]
Comment (vr009) [reply cli/pack/tgz.go]: Fixed.

File: cli/pack/tgz.go [hunk 39]
```diff
@@ -81,6 +81,39 @@ func WriteTarArchive(srcDirPath string, compressWriter io.Writer) error {
...
+	destFile, err := os.Create(destFilePath)
+	if err != nil {
+		return fmt.Errorf("Failed to create result GZIP file %s: %s", destFilePath, err)
+	}
+	defer destFile.Close()
+
+	// dest file GZIP writer
```
[the remark this answers is not in the document]
Comment (vr009) [reply cli/pack/tgz.go]: Fixed.

File: cli/pack/tgz.go [hunk 40]
```diff
@@ -81,6 +81,39 @@ func WriteTarArchive(srcDirPath string, compressWriter io.Writer) error {
...
+	gzipWriter, err := gzip.NewWriterLevel(destFile, gzip.BestCompression)
+	if err != nil {
+		return fmt.Errorf("Failed to create GZIP writer %s: %s", destFilePath, err)
+	}
+	defer gzipWriter.Flush()
+
+	// compressing itself
```
[the remark this answers is not in the document]
Comment (vr009) [reply cli/pack/tgz.go]: Fixed.

File: cli/pack/tgz.go [hunk 41]
```diff
@@ -81,6 +81,39 @@ func WriteTarArchive(srcDirPath string, compressWriter io.Writer) error {
...
+	if _, err := io.Copy(gzipWriter, srcFileReader); err != nil {
+		return err
```
[the remark this answers is not in the document]
Comment (vr009) [reply cli/pack/tgz.go:111]: Fixed.

File: cli/util/crypto.go [hunk 42]
```diff
@@ -0,0 +1,81 @@
...
+	hasher := sha1.New()
+	if _, err := io.Copy(hasher, file); err != nil {
+		return "", err
+	}
+
+	return fmt.Sprintf("%x", hasher.Sum(nil)), nil
+}
+
+// FileMD5 computes MD5 for a given file.
+// The result is returned in a binary form
```
[the remark this answers is not in the document]
Comment (vr009) [reply cli/util/crypto.go]: Fixed.

File: cli/util/crypto.go [hunk 43]
```diff
@@ -0,0 +1,81 @@
...
+	hasher := sha256.New()
+	if _, err := io.Copy(hasher, file); err != nil {
+		return "", err
+	}
+
+	return fmt.Sprintf("%x", hasher.Sum(nil)), nil
+}
+
+// FileSHA1Hex computes SHA1 for a given file.
+// The result is returned in a hex form
```
[the remark this answers is not in the document]
Comment (vr009) [reply cli/util/crypto.go]: Fixed.

File: cli/util/crypto.go [hunk 44]
```diff
@@ -0,0 +1,81 @@
...
+	hasher := md5.New()
+	if _, err := io.Copy(hasher, file); err != nil {
+		return nil, err
+	}
+
+	return hasher.Sum(nil), nil
+}
+
+// FileMD5Hex computes MD5 for a given file.
+// The result is returned in a hex form
```
[the remark this answers is not in the document]
Comment (vr009) [reply cli/util/crypto.go]: Fixed.

File: cli/util/util.go [hunk 45]
```diff
@@ -746,3 +747,39 @@ func GetAbsPath(baseDir, filePath string) string {
 	}
 	return filepath.Join(baseDir, filePath)
 }
+
+// ConcatBuffers appends sources content to dest
```
[the remark this answers is not in the document]
Comment (vr009) [reply cli/util/util.go]: Fixed.

File: cli/util/util.go [hunk 46]
```diff
@@ -746,3 +747,39 @@ func GetAbsPath(baseDir, filePath string) string {
...
+	for _, src := range sources {
+		if _, err := io.Copy(dest, src); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// MergeFiles creates a file that is a concatenation of srcFilePaths
```
[the remark this answers is not in the document]
Comment (vr009) [reply cli/util/util.go]: Fixed.

File: cli/util/crypto.go [hunk 47]
```diff
@@ -0,0 +1,81 @@
...
+	fileMD5, err := FileMD5(path)
+	if err != nil {
+		return "", err
+	}
+
+	return fmt.Sprintf("%x", fileMD5), nil
+}
+
+// StringSHA1Hex computes SHA1 for a given string
+// The result is returned in a hex form
```
[the remark this answers is not in the document]
Comment (vr009) [reply cli/util/crypto.go]: Fixed.

File: cli/pack/rpm_pack.go [hunk 48]
```diff
@@ -0,0 +1,218 @@
...
+
+	// create body file = header + compressedCpio
+	rpmBodyFilePath := filepath.Join(packageDir, "body")
+	if err := util.MergeFiles(rpmBodyFilePath, rpmHeaderFilePath, compresedCpioPath); err != nil {
+		return fmt.Errorf("Failed to concat RPM header with compressed payload: %s", err)
+	}
+
+	log.Infof("Computing a signature")
+
+	// compute signature
```
[the remark this answers is not in the document]
Comment (vr009) [reply cli/pack/rpm_pack.go]: Fixed.

File: cli/pack/rpm_tagset_test.go [hunk 49]
```diff
@@ -0,0 +1,194 @@
...
+	assert.Equal("000000010000000200000064ffffffbe", hex(packed.Data))
+
+	// INT64
+	int64Tag := rpmTagType{
+		Type:  rpmTypeInt64,
+		Value: []int64{1, 2, 100, -66},
+	}
+
+	packed, err = packTag(int64Tag)
+	assert.Nil(err)
```
[the remark this answers is not in the document]
Comment (vr009) [reply cli/pack/rpm_tagset_test.go:123]: Added.

File: cli/pack/rpm_tagset_test.go [hunk 50]
```diff
@@ -0,0 +1,194 @@
...
+
+	// tags set with padding
+	data, err = packTagSet(rpmTagSetType{
+		// 5 bytes string (4 symbols + NULL)
+		{ID: tagName, Type: rpmTypeString, Value: "abcd"},
+		// int32 value should be aligned on 4-byte boundaries
+		{ID: tagDirIndexes, Type: rpmTypeInt32, Value: []int32{1, 2, 3}},
+	}, regionTagID)
+
+	assert.Nil(err)
```
[the remark this answers is not in the document]
Comment (vr009) [reply cli/pack/rpm_tagset_test.go:240]: Added.

File: cli/pack/rpm_header.go [hunk 51]
```diff
@@ -0,0 +1,269 @@
...
+		filesInfo.FileModes = append(filesInfo.FileModes, int16(sysFileInfo.Mode))
+		filesInfo.FileInodes = append(filesInfo.FileInodes, int32(sysFileInfo.Ino))
+		filesInfo.FileDevices = append(filesInfo.FileDevices, int32(sysFileInfo.Dev))
+		filesInfo.FileRdevs = append(filesInfo.FileRdevs, int16(sysFileInfo.Rdev))
+	}
+
+	return filesInfo, nil
+}
+
+func addDirAndGetIndex(dirNames *[]string, fileDir string) int {
```
[the remark this answers is not in the document]
Comment (vr009) [reply cli/pack/rpm_header.go:254]: Added.

File: cli/pack/rpm_header.go [hunk 52]
```diff
@@ -0,0 +1,269 @@
...
+		if fileInfo.Mode().IsRegular() {
+			filesInfo.FileFlags = append(filesInfo.FileFlags, fileFlag) // XXX
+
+			fileDigest, err := util.FileMD5Hex(fullFilePath)
+			if err != nil {
+				return filesInfo, fmt.Errorf("Failed to get file MD5 hex: %s", err)
+			}
+
+			filesInfo.FileDigests = append(filesInfo.FileDigests, fileDigest)
+		} else {
```
[the remark this answers is not in the document]
Comment (vr009) [reply cli/pack/rpm_header.go:218]: No need, there is already a such check above.

File: cli/pack/rpm_header.go [hunk 53]
```diff
@@ -0,0 +1,269 @@
...
+	FileDevices    []int32
+	FileMtimes     []int32
+	FileLangs      []string
+	FileRdevs      []int16
+	FileLinkTos    []string
+	FileFlags      []int32
+	FileDigests    []string
+}
+
+func getRPMRelation(relation string) int32 {
```
[the remark this answers is not in the document]
Comment (vr009) [reply cli/pack/rpm_header.go:37]: Added.

File: cli/pack/deb.go (hunk 9 above)
Comment (vr009) [reply cli/pack/deb.go]: Fixed.

File: cli/pack/rpm_tagset.go [hunk 54]
```diff
@@ -0,0 +1,390 @@
...
+ *   `tagSetHeader` + `resIndex` + `resData`
+ *
+ */
+
+// packTagSet packs all passed tags into a buffer and returns it.
+func packTagSet(tagSet rpmTagSetType, regionTagID int) (*bytes.Buffer, error) {
+	var tagsData = bytes.NewBuffer(nil)
+	var tagsIndex = bytes.NewBuffer(nil)
+
+	// tagsData and tagsIndex
```
[the remark this answers is not in the document]
Comment (vr009) [reply cli/pack/rpm_tagset.go:186]: The full explanation is above at the top of the document, these notes are for some kind of navigation here.

File: cli/pack/rpm_tagset.go [hunk 55]
```diff
@@ -0,0 +1,390 @@
...
+	tagsNum := len(tagSet) + 1
+	regionTagData := getPackedTagIndex(regionTagID, rpmTypeBin, -tagsNum*16, 16)
+
+	// resData is tagsData + regionTagData
+	var resData = bytes.NewBuffer(nil)
+	if err := util.ConcatBuffers(resData, tagsData, regionTagData); err != nil {
+		return nil, err
+	}
+
+	// tagSetHeader
```
[the remark this answers is not in the document]
Comment (vr009) [reply cli/pack/rpm_tagset.go:229]: The full explanation is above at the top of the document, these notes are for some kind of navigation here.

File: cli/pack/common.go [hunk 56]
```diff
@@ -502,40 +502,86 @@ func getVersion(packCtx *cmdcontext.PackCtx) string {
...
+		ttBin = realPath
+	}
+
+	err = copy.Copy(ttBin, filepath.Join(destPath, filepath.Base(ttBin)))
+	if err != nil {
+		return err
+	}
+
+	if cmdCtx.Cli.IsSystem {
+		cmd := exec.Command("which", "tarantool")
```
[the remark this answers is not in the document]
Comment (vr009) [reply cli/pack/common.go]: Is not it valid if a user set flags --with-binaries and uses system env?

[the remark this answers is not in the document]
Comment (psergee) [reply cli/pack/common.go]: I don't think these flags affect the way we search for binaries.

[the remark this answers is not in the document]
Comment (vr009) [reply cli/pack/common.go]: Fixed.

**Review by psergee — APPROVED** [review]

**Pull request merged**